### DISCUSSION TITLE
Help For a Totally Confused Newbie

### DISCUSSION THREAD

**Discussion by PeteCal** [zwave-js/zwave-js-ui#2507] [Q&A] [answered]
First, I should probably RTFM but I can't find the FM.  
Second, after a week or more, I have zwavejs2mqtt installed on my RPi. The Control Panel comes up and I must have gotten things right because it has found four of my six z-wave devices.  It can't find my two First Alert CO detectors but they have always been a problem.
I don't want to use Home Assistant.  I have my own network of door sensors, window sensors, smoke detectors, glass break sensors, garage door controller, etc.  They all communicate with mqtt.  
I thought I would be able to mosquitto_pub and mosquitto_sub to the zwave devices.  But after many hours of searching and thrashing, I can't figure out how.
Note, I have been using my network with an open source "bus" but I need to get away from that.

Any help would be greatly appreciated.

Pete

**Comment by kpine** [comment]
> First, I should probably RTFM but I can't find the FM.

FYI, it's linked from the [landing page](https://github.com/zwave-js/zwavejs2mqtt) of this repository.

![image](https://user-images.githubusercontent.com/791794/177020397-7363c21b-067d-4ada-8e60-a4e2cbe366b0.png)

![image](https://user-images.githubusercontent.com/791794/177020401-e970b9bc-e0ad-42ed-9630-f25159946898.png)

In the documentation, there's a section [Using MQTT](https://zwave-js.github.io/zwavejs2mqtt/#/guide/mqtt) that describes the MQTT API.

**Comment by PeteCal** [comment]
kpine,
For the last few days, I have been all over that page and similar pages.  I don't know what to do with that info.  Maybe it is because I know nothing about JSON.  I don't know.  For example, what do I do with 
"<mqtt_prefix>/_EVENTS_/ZWAVE_GATEWAY-<mqtt_name>/<driver|node|controller>/<event_name>"
Is that something I type on a terminal line?  I doubt it.  Nothing on those pages mean anything to me.

I am concerned that I have a lot to learn and don't know where to start.

Thanks for your help.

Pete

**Reply by kpine** [reply]
That is an example of an MQTT topic. What you do with it is up to you. For that particular example, maybe it's not something you care about...

I would suggest using [MQTT Explorer](https://mqtt-explorer.com/) to get a better idea of the topic formatting and the payloads.

**Reply by kpine** [reply]
Setting values (via `mosquitto_pub`, for example), would be done with the [Set Value API](https://zwave-js.github.io/zwavejs2mqtt/#/guide/mqtt?id=set-values). MQTT Explorer will help you determine which values to set.

In the [Gateway configuration](https://zwave-js.github.io/zwavejs2mqtt/#/usage/setup?id=gateway), you specify how you want the topics to be formatted. It has examples of the payload types.

**Comment by PeteCal** [comment]
`kpine,`
Well, I got MQTTExplorer working on my windows computer.  I wanted it on the RPi but ....long story.

As it stands now, MQTTE makes sense to me for all my devices that use mqtt.  MQTTE displays all my sensor reports and I can use it to command my garage door to open by using the "Publish" button.

But it doesn't show any of the activity on the zwave side.  Maybe because zwavejs2mqtt isn't set up correctly.  I am exploring that.

I have also been working with one of my zwave devices, a thermostat, that I have been using for years and I am very familiar with.  But I still haven't figured out how to "say",:
"mosquitto_pub -t ThermostatUS -m 'Cooling' " or something to that effect.  I see in Gateway setup the format is:

`<mqtt_prefix>/<?node_location>/<nodeId>/<commandClass>/<endpoint>/<property>/<propertyKey>`

And I have concluded that:

`<mqtt_prefix>` = zwave
`<?node_location>` = Second Floor
`<nodeId>` = 5
`<commandClass>=` 66
`<endpoint>=` I have no idea
`<property>=2`  (that is the value for "cooling")
`<propertyKey>=I` have no idea.


But I don't know what to do with that information.  How do I turn it into something mosquitto can publish?


I know I am missing something very important but I haven't cracked that nut yet.

**Reply by robertsLando** [reply]
This needs some mqtt basis an also zwave basis, by default the root endpoint (the device itself) is considered 0, if the device has (for example) multiple temperature sensor or multiple outputs it will map them on other endpoints (1, 2, 3 etc..). Propertykey is not always defined, it is for values like colors to allow mappping all color components R-G-B or R-G-B-A  to a different valueId, also known as [partial parameters](https://zwave-js.github.io/node-zwave-js/#/config-files/file-format?id=partial-parameters)

**Comment by PeteCal** [comment]
robertsLando,
Thanks.  But I have to say that link means absolutly nothing to me.  It has to do with Z-Wave JS.  I know nothing about Z-Wave JS or javascript.  I got here by trying to use zwave2mqtt.  I can't get anywhere.  I have spent 30 or 40 hours trying to create:
mosquitto_pub -t???? -m???? 
with zero success.  I am no closer to being able to do that then I was at first start.
I think I am going to have to put zwave2mqtt in a Time Out and look for another solution.  I may have to come back if that fails, but I cring at the thought.
Thanks, and thanks to kpine.

**Reply by robertsLando** [reply]
Here is how I do it for testing:
https://zwave-js.github.io/zwavejs2mqtt/#/guide/mqtt?id=example-setting-a-value

**Comment by PeteCal** [comment]
I had actually tried to follow that video with my system a few days ago.  I couldn't get anything on MQTTExplorer.   I tried quite a few changes to my configuration/setup but couldn't get it to work.  I finally considered that my system was differient enough that it wouldn't work.  MQTTExplorer never saw any messages from zwave2mqtt even though zwave2mqtt could do things like change the set point temperature.  Also, MQTTExplorer sees all of my other mqtt pubs.

Given your recommendation I tried again.  I spent a couple of hours reviewing the definations of the various settings and tried more variations but no luck.  
I am still getting, from the log the following, repeated every 3 seconds,

Error: connect ECONNREFUSED 127.0.0.1:1883
at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1133:16)
2022-07-06 18:13:27.439 INFO MQTT: MQTT client closed
2022-07-06 18:13:30.438 INFO MQTT: MQTT client reconnecting
2022-07-06 18:13:30.445 ERROR MQTT: Mqtt client error connect ECONNREFUSED 127.0.0.1:1883
Error: connect ECONNREFUSED 127.0.0.1:1883

**Reply by robertsLando** [reply]
@PeteCal Do you have an mqtt broker running on your machine? Mosquitto for example? 

http://www.steves-internet-guide.com/install-mosquitto-linux/

Once that is running you will see messages coming

**Comment by PeteCal** [comment, marked as answer]
Yes, I have mosquitto running.  It supports 10 or 15 IOT devices in my system.  MQTTExplorer catches all the pubs and displays them correctly.  My problem WAS...I say WAS because I found it last night.  Just like I said I was missing something important.  And, actually what I was missing was the most basic...lets say stupid thing.  When I changed the MQTT Host URL, I failed to hit the ENTER key.  So when I went off that page it reverted to localhost.  I am running zwavejs2mqtt in a docker container so I needed to change that to the URL of my Raspberry Pi even though it is the same computer.  Mosquitto is not in the container with zwavejs2mqtt so localhost won't work.  I don't know how many times, while trying to get this to work, I went back checking my settings and discovered localhost had replaced 192.168.1.115 (my RPi URL) over and over.  That was because I didn't hit ENTER every time I typed in the URL.  It can't get more mundane then that.
Thanks 
RobertLando and kpine

**Reply by robertsLando** [reply]
Glad you fixed it :)
